src/modbus-tcp-server.c

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright © 2016 DEIF A/S
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#ifndef _MSC_VER
+# include <unistd.h>
+#endif
+#include <time.h>
+
+#if defined(_WIN32)
+# define OS_WIN32
+/* ws2_32.dll has getaddrinfo and freeaddrinfo on Windows XP and later.
+ * minwg32 headers check WINVER before allowing the use of these */
+# ifndef WINVER
+# define WINVER 0x0501
+# endif
+/* Already set in modbus-tcp.h but it seems order matters in VS2005 */
+# include <winsock2.h>
+# include <ws2tcpip.h>
+# define SHUT_RDWR 2
+# define TIME_UTC 1
+# define close closesocket
+#else
+# include <sys/select.h>
+# include <sys/socket.h>
+# include <netinet/in.h>
+# include <arpa/inet.h>
+# include <netdb.h>
+#endif
+
+#include "modbus.h"
+#include "modbus-private.h"
+#include "modbus-tcp-server.h"
+#include <pthread.h>
+#include <time.h>
+
+#include <modbus.h>
+
+#if defined (_WIN32)
+# define CLOCK_MONOTONIC 1
+# ifndef HAVE_STRUCT_TIMESPEC
+#  define HAVE_STRUCT_TIMESPEC 1
+struct timespec {
+        long tv_sec;
+        long tv_nsec;
+};
+#endif /* HAVE_STRUCT_TIMESPEC */
+
+static int clock_gettime(int clock_type, struct timespec *tv)
+{
+   tv->tv_sec = time(NULL);
+   tv->tv_nsec = 0;
+   return 0;
+}
+#endif
+
+#define MB_TCP_SRV_IDLE_TIMEOUT 60
+
+/* mdbus client data */
+struct modbus_tcp_client_t
+{
+   int socket;
+   time_t last_update;
+   struct modbus_tcp_client_t* next;
+};
+
+/* mdbus server data */
+struct _modbus_tcp_server
+{
+   int server_socket;
+   fd_set refset;
+   fd_set rdset;
+   int16_t fdmax;
+   uint16_t max_connections;
+   uint16_t port;
+   uint32_t idle_time_sec;
+   uint32_t select_to_sec;
+   uint32_t select_to_usec;
+   uint8_t shutdown;
```

Comment:
Since libmodbus is very crossplatform and doesn't use bool anywhere, I am a bit 'scared' to add it.